Bricks/sync/owned_borrowed.h

```diff
@@ -0,0 +1,492 @@
+    rhs.p_actual_instance_ = nullptr;
+  }
+
+  WeakBorrowed(const WeakBorrowed&) = delete;
+  WeakBorrowed(WeakBorrowed&&) = delete;
+  WeakBorrowed& operator=(const WeakBorrowed&) = delete;
+  WeakBorrowed& operator=(WeakBorrowed&&) = delete;
+
+ public:
+  ~WeakBorrowed() { InternalUnRegister(); }
+
+  // Whether the object is valid and has not yet been marked for destruction.
+  // The only way to invalidate the object is to move away from it.
+  // The only way to mark it for destruction is for the `Owned` object to get out of scope.
+  // THREAD-SAFE. NEVER THROWS.
+  bool IsValid() const { return p_actual_instance_ && !p_actual_instance_->destructing_; }
+  operator bool() const { return IsValid(); }
+
+  // Smart-pointer-style accessors.
+  // THREAD-SAFE. NEVER THROWS.
+  T& operator*() { return p_actual_instance_->instance_; }
+  const T& operator*() const { return p_actual_instance_->instance_; }
+  T* operator->() { return &p_actual_instance_->instance_; }
+  const T* operator->() const { return &p_actual_instance_->instance_; }
+
+  // Exclusive accessors.
+  // THREAD-SAFE WITH RESPECT TO MAKING SURE UNDERLYING OBJECT IS ACCESSED EXCLUSIVELY BY THE PASSED IN FUNCTOR.
+  // NEVER THROWS.
+  void ExclusiveUse(std::function<void(T&)> f) {
+    std::lock_guard<std::mutex> lock(p_actual_instance_->mutex_);
+    f(operator*());
+  }
+  void ExclusiveUse(std::function<void(const T&)> f) const {
+    std::lock_guard<std::mutex> lock(p_actual_instance_->mutex_);
+    f(operator*());
+  }
+  template <typename V>
+  V ExclusiveUse(std::function<V(T&)> f) {
+    std::lock_guard<std::mutex> lock(p_actual_instance_->mutex_);
+    return f(operator*());
+  }
+  template <typename V>
+  V ExclusiveUse(std::function<V(const T&)> f) const {
+    std::lock_guard<std::mutex> lock(p_actual_instance_->mutex_);
+    return f(operator*());
+  }
+
+  // Return the number of registered borrower users besides the master one.
+  // For unit-testing purposes mostly, but may end up useful. -- D.K.
+  // THREAD-SAFE. NEVER THROWS.
+  size_t NumberOfActiveBorrowers() const {
+    std::lock_guard<std::mutex> lock(p_actual_instance_->mutex_);
+    return p_actual_instance_->borrowers_.size();
+  }
+
+  // Return the total number of registered borrower users registered, with some possibly already out of scope.
+  // For unit-testing purposes mostly, but may end up useful. -- D.K.
+  // THREAD-SAFE. NEVER THROWS.
+  size_t TotalBorrowersSpawnedThroughoutLifetime() const {
+    std::lock_guard<std::mutex> lock(p_actual_instance_->mutex_);
+    return p_actual_instance_->total_borrowers_spawned_throughout_lifetime_;
+  }
+
+ protected:
+  instance_t* p_actual_instance_;  // Is set to `nullptr` externally when moved.
+
+  void InternalUnRegister() {
+    if (key_) {
+      CURRENT_ASSERT(p_actual_instance_);
+      // `*this` is a valid (non - std::move()-d - away) `BorrowedWithCallback`.
```

Comment:
This function is also called by `Borrower`, so looks like this is not always true.